OpenRA.Mods.Common/Traits/Crates/GrantExternalConditionCrateAction.cs

```diff
@@ -70,11 +73,23 @@ public override void Activate(Actor collector)
 					if (!a.IsInWorld || a.IsDead)
 						continue;
 
-					var external = a.TraitsImplementing<ExternalCondition>()
-						.FirstOrDefault(t => t.Info.Condition == info.Condition && t.CanGrantCondition(a, self));
+					var externals = a.TraitsImplementing<ExternalCondition>().Where(t => t.Info.Condition == info.Condition);
```

Comment:
This can simplify a bit further to reduce duplication:
```csharp
var externals = a.TraitsImplementing<ExternalCondition>()
	.Where(t => t.Info.Condition == info.Condition);

ExternalCondition external = null;
for (var n = 0; n < info.Levels; n++)
{
	if (external == null || !external.CanGrantCondition(a, self))
	{
		external = externals.FirstOrDefault(t => t.CanGrantCondition(a, self));
		if (external == null)
			break;
	}

	external.GrantCondition(a, self, info.Duration);
}
```